### PR TITLE
[WIP] Allow objects parameters in mutations

### PR DESCRIPTION
**Pull request by prevostc**
This is a work in progress to add support for object parameters in mutations and thus support reindex.io default mutation format.

I did succeed in making all tests pass BUT the fix is obviously wrong so I reach out here to get feedback on which other part of the project should be breaking so I can add some tests there and create a proper fix.

I also reused a duplicated test.

### CODE REVIEW

**Comment by stubailo** [comment]
I think the way to do this is to use `valueFromAST` from the `graphql` utils: https://github.com/graphql/graphql-js/blob/master/src/utilities/valueFromAST.js

Basically, what you're doing here is using the AST nodes for the object in the store, but we need to use the actual values.

**Comment by prevostc** [comment]
All right, I did find how to use that function but I can't find a way to get the right `GraphQLType`. It seems like I need a `GraphQLSchema` to call `getType` on it but I'm not sure.

**Comment by helfer** [comment]
@prevostc Since the goal of your PR is to support objects as inputs to mutations, I think what you're looking for is `GraphQLInputObjectType`. If that's the case, you won't need a schema, because this is a generic type that you can import from GraphQL.

**Comment by prevostc** [comment]
@helfer sorry I didn't explained the whole story, I do need a `GraphQLInputObjectType` as the second argument of  `valueFromAST` but I may need an instance of it because if I try to call it with `valueFromAST(value, GraphQLInputObjectType)` I get this compilation error:
```
Argument of type 'typeof GraphQLInputObjectType' is not assignable to parameter of type 'GraphQLScalarType | GraphQLEnumType | GraphQLInputObjectType | GraphQLList | GraphQLNonNull'.
```

I did some research and the `GraphQLSchema` has a way to retrieve defined types, but this might not be the right way to get the corresponding graphql type.

**Comment by helfer** [comment]
Oh, yes, of course you need an instance.

Getting the type from the schema works just fine if you have the schema, but we're trying not to require having the schema on the client, because it means either doing an extra round-trip, or making the build process more complicated.

Since in Apollo client mutations with input objects as variables currently don't do any validation, it probably makes most sense to do the same with inline input objects, i.e. just parse the thing without bothering to check the types at all.

To achieve that, you can copy the `valueFromAST` function and modify it so it doesn't require instances for `InputObjectType` and `ListType`. Then you just call `modifiedValueFromAST(astValue, variables)` without having to pass the type.

Does that make sense?

**Comment by prevostc** [comment]
Thanks @helfer for the guidance. After some heavy code reading, I did not copied the `valueFromAST` because it relied on  `type.parseLiteral` and as you mentioned I don't have a type object to work with. 

Instead I reworked the existing `storeKeyNameFromField` code to be called recursively on `ObjectValue` like `valueFromAST` do for nested objects.

I'm not sure that this new function should be named `valueFromAST` though as the signature is not the same on both packages, this might be confusing.

**Comment by helfer** [comment]
That's great! `valueFromAST` is definitely not the right name, as it's already taken. Just give it another name that describes what it does.

**Comment by prevostc** [comment]
@helfer done \o/ Should I do something else ?

**Comment by stubailo** [comment]
@prevostc one thing you should do is sign the CLA - it lets us make sure that people are contributing code that they own.

File: typings/browser/globals/graphql/index.d.ts [hunk 1]
```diff
@@ -936,4 +936,7 @@ declare module "graphql" {
         directives?: Array<GraphQLDirective>;
     }
 
-}
\ No newline at end of file
+// utilities/valueFromAST.js
```
Comment (stubailo) [inline typings/browser/globals/graphql/index.d.ts]: We don't need this, do we?

**Comment by stubailo** [comment]
I would really prefer to have a few more tests here though - just changing that one doesn't really cover a lot of space of possible arguments. We could just add tests for `valueToObjectRepresentation` directly, without running any other code.

File: typings/browser/globals/graphql/index.d.ts (hunk 1 above)
Comment (prevostc) [inline typings/browser/globals/graphql/index.d.ts]: Without this declaration, the ts compilation fails. As for the comment line, I tried to use the same convention already present in the file that is to have a section prefixed with the corresponding file name.

**Comment by helfer** [comment]
@prevostc Any update on this? I'd like to merge it as soon as there are some more tests and you've signed the CLA.

**Comment by prevostc** [comment]
@helfer I will be able to add more tests by Monday. I will add more Object to string representation tests, including all scalar types, 2 or 3 levels of nested object, arrays, arrays of objects and edge cases like nulls and cycling references.

**Comment by prevostc** [comment]
@helfer Also, I  did sign the CLA, but it keeps saying that I didn't :(

**Comment by helfer** [comment]
@prevostc Perfect, I think it shows up as signed now.

I don't think you need to test cycles, since the AST gets parsed from a query string, which cannot possibly have any cycles. Everything else sounds good. 2 levels of nesting should be enough.

File: src/data/storeUtils.ts [hunk 2]
```diff
@@ -24,26 +27,47 @@ function isVariable(value: Value): value is Variable {
...
+    const variableValue = variables[value.name.value];
+    argObj[name.value] = variableValue;
+  } else if (isList(value)) {
+    argObj[name.value] = value.values.map((listValue) => {
+      const nestedArgArrayObj = {};
+      valueToObjectRepresentation(nestedArgArrayObj, name, listValue, variables);
+      return nestedArgArrayObj[name.value];
+    });
+  } else {
+    throw new Error(`For inline arguments, only scalar and Object types are supported.
```
Comment (stubailo) [inline src/data/storeUtils.ts]: Probably this should just say enum arguments are not supported, since I think we have covered literally everything else: http://facebook.github.io/graphql/#Value

Comment (prevostc) [inline src/data/storeUtils.ts]: I added Enum support in the next commit and reworked the error message to be more generic.

File: src/data/storeUtils.ts [hunk 3]
```diff
@@ -35,8 +36,12 @@ function isList(value: Value): value is ListValue {
   return value.kind === 'ListValue';
 }
 
+function isEnum(value: Value): value is EnumValue {
+  return value.kind === 'EnumValue';
+}
+
 function valueToObjectRepresentation(argObj: Object, name: Name, value: Value, variables?: Object) {
-  if (isScalarValue(value)) {
+  if (isScalarValue(value) || isEnum(value)) {
```
Comment (stubailo) [inline src/data/storeUtils.ts]: This is not correct - I think we shouldn't support enums inline because then if you do the same query with variables and with an enum inline, they won't be cached the same way. It's not really possible to know the underlying value of an enum entry on the client.

**Comment by stubailo** [comment]
I'd request that we add tests that ensure that arguments are serialized the same way if they are passed via inline AST or variables.

File: src/data/storeUtils.ts (hunk 3 above)
Comment (prevostc) [inline src/data/storeUtils.ts]: ok, I will use isEnum to show a detailed error message then.

**Comment by prevostc** [comment]
@stubailo only numbers are not handled the same way, integers are not cast to numbers when inlined. fixing this broke some tests (see below), I don't know what's the right fix here.
```1) writing to the store properly normalizes a aliased fields with arguments:

      AssertionError: expected { Object (abcd) } to deeply equal { Object (abcd) }
      + expected - actual

         "abcd": {
           "id": "abcd"
           "nullField": [null]
           "numberField": 5
      -    "stringField({"arg":1})": "The arg was 1!"
      -    "stringField({"arg":2})": "The arg was 2!"
      +    "stringField({"arg":"1"})": "The arg was 1!"
      +    "stringField({"arg":"2"})": "The arg was 2!"
         }
       }
```
```2) client can allow the store to be rehydrated from the server:

      AssertionError: expected { Object (apollo) } to deeply equal { Object (apollo) }
      + expected - actual

         "apollo": {
           "data": {
             "ROOT_QUERY": {
               "allPeople({"first":"1"})": "ROOT_QUERY.allPeople({\"first\":\"1\"})"
      +        "allPeople({"first":1})": "ROOT_QUERY.allPeople({\"first\":1})"
             }```

**Comment by stubailo** [comment]
Hmm I think we should make integers always cast to numbers and update the tests. Strings in graphql literals always have quotes, so it shouldn't be ambiguous I think!

**Comment by stubailo** [comment]
By the way, thanks for putting so much time into the PR, this is really great stuff.

File: test/QueryManager.ts [hunk 4]
```diff
@@ -979,10 +979,10 @@ describe('QueryManager', () => {
     });
   });
 
-  it('runs a mutation and puts the result in the store', () => {
+  it('runs a mutation with object parameters and puts the result in the store', () => {
```
Comment (helfer) [inline test/QueryManager.ts]: It's not clear what exactly this test is supposed to do. Originally it was to test that the result of a mutation is properly put in the store. We should keep the old test and create an additional one for object values.

**Comment by helfer** [comment]
@prevostc This is shaping up very nicely, thanks a lot for all your work!

My only nitpick is with the tests. Currently, there's just one unit large test that presumably covers all of your changes. If this test fails in the future, it will be very hard to determine what exactly isn't working. Instead of writing one big test, we generally try to test each bit of functionality separately with a smaller test.

So you should write a separate test for each branch of the `valueToObjectRepresentation` function:

- for a scalar
- for a number
- for an object
- for a variable
- for a missing variable
- for a list
- for an unknown kind (making sure the correct error is thrown)

You can do that by importing the `valueToObjectRepresentation` function directly, and then making sure the output is what it should be.

The big test you wrote is more like an integration test, so it's a perfectly good test that you should keep, but it should run after the smaller unit tests for the `valueToObjectRepresentation`. That way it will be much easier to know what's broken when a test fails.

File: test/QueryManager.ts (hunk 4 above)
Comment (prevostc) [inline test/QueryManager.ts]: This test was a duplicate from the test at line 938, I just reused it instead of dropping it and recreating it.

Comment (helfer) [inline test/QueryManager.ts]: Ah, yes, I didn't see that.

**Comment by helfer** [comment]
@prevostc Just give me a nudge when you think it's ready to be reviewed again, ok?

**Comment by prevostc** [comment]
@helfer I added a unit test that cover each case individually. I did the "data provider" manually, maybe there is a best way :s

**Comment by helfer** [comment]
Ah, ok, I wasn't sure if you were still working on it or not, so I didn't take a look yet. It will have to be rebased before we can merge, but I can take a look first.

**Comment by helfer** [comment]
@prevostc The new tests look great! 👍  This is 99% ready to merge, just some tiny things to fix.

I took a look at the coverage report, and it seems there are still two lines that aren't tested at all. You can see them here: https://coveralls.io/builds/6534889/source?filename=src%2Fdata%2FstoreUtils.ts

Can you add two tests that cover these lines? One is for a missing variable, the other is for unsupported inline type (just enum, I believe).

Also, please rebase your PR so it can be merged without conflict. If you're not sure how to do that, just message me on the apollo slack.

**Comment by stubailo** [comment]
This PR is amazing.

**Comment by prevostc** [comment]
@helfer Done :) I also squashed all commits so you can keep a clean commit history too.

**Comment by helfer** [comment]
Awesome, this is perfect now. Merged!

**Comment by prevostc** [comment]
Oups, sry. Thanks for you careful reviews!

**Comment by helfer** [comment]
No worries, thanks a lot for the PR, it's really great! It would be really cool if you want to be an active contributor! I can add you to our team, so you can work on branches directly on this repo. How does that sound?

**Comment by danielbuechele** [comment]
I am wondring how the suggested workaround with variables was working before this PR. I have the code below, but I am getting an error.
```
mutation: gql`
	mutation(
		$title: String!,
		$status: ItemStatus!,
	) {
		item(item: {
			title: $title,
			status: $status,
		}) {
			id
		}
	}
`,
variables: {
	title: raw.title,
	status: raw.status,
},
```

**Comment by stubailo** [comment]
> I am getting an error.

What does the error say?

**Comment by danielbuechele** [comment]
`VM4646:263 Uncaught Error: For inline arguments, only scalar types are supported. To use Enum or Object types, please pass them as variables.`
I am using `apollo-client@0.3.13`

**Comment by prevostc** [comment]
This issue have been fixed by the merged PR but as a quickfix you could pass the `item` field as a variable (might not work):

```
mutation: gql`
    mutation(
        $title: String!,
        $status: ItemStatus!,
    ) {
        item(item: $item) {
            id
        }
    }
`,
variables: {
    item: {
        title: raw.title,
        status: raw.status,
    }
},
```

**Comment by stubailo** [comment]
@prevostc has the right workaround. Since `item` is an object, you had to pass _the whole object_ as a variable.
